File: Codes/Programmers/압축.py
def solution(msg):
    answer = []
    word = {chr(i + 64): i for i in range(1, 27)}
    now_index = 27

    i = 0
    now_word = ""
    while i < len(msg):
        nxt_word = now_word + msg[i]
        if nxt_word not in word:
            answer.append(word[now_word])
            word[nxt_word] = now_index
            now_index += 1
            now_word = ""
        else:
            now_word = nxt_word
            i += 1

    answer.append(word[now_word])

    return answer

File: Codes/Programmers/압축.js
function solution(msg) {
  const answer = [];
  const d = {};
  for (let i = 1; i < 27; i++) {
    d[String.fromCharCode(i + 64)] = i;
  }

  let left,
    right = 0;
  const n = msg.length;
  let m = 26;

  while (right < n) {
    let now = msg.slice(left, right + 1);
    if (!(now in d)) {
      d[now] = ++m;
      answer.push(d[msg.slice(left, right)]);
      left = right;
    } else {
      right++;
    }
  }
  answer.push(d[msg.slice(left, right + 1)]);

  return answer;
}
